src/sms.rs

```diff
@@ -0,0 +1,201 @@
+    0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77, /* 112-119: p q r s t u v w  */
+    0x78, 0x79, 0x7A, 0xA8, 0xC0, 0xA9, 0xBD, 0x20, /* 120-127: x y z { | } ~ DEL */
+    /* Character codes 128-255 (beyond standard ASCII) have different possible
+     * interpretations. This table has been done according to ISO-8859-15.
+     */
+    0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, /* 128-159: Undefined   */
+    0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20,
+    0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x40, 0x63, 0x01, 0xE5, 0x03, 0x53,
+    0x5F, /* 160-167: ..£, €... */
+    0x73, 0x63, 0x20, 0x20, 0x20, 0x2D, 0x20, 0x20, /* 168-175 */
+    0x20, 0x20, 0x20, 0x20, 0x5A, 0x75, 0x0A, 0x20, /* 176-183 */
+    0x7A, 0x20, 0x20, 0x20, 0x20, 0x20, 0x59, 0x60, /* 184-191 */
+    0x41, 0x41, 0x41, 0x41, 0x5B, 0x0E, 0x1C, 0x09, /* 192-199: ..Ä, Å... */
+    0x45, 0x1F, 0x45, 0x45, 0x49, 0x49, 0x49, 0x49, /* 200-207 */
+    0x44, 0x5D, 0x4F, 0x4F, 0x4F, 0x4F, 0x5C, 0x2A, /* 208-215: ..Ö... */
+    0x0B, 0x55, 0x55, 0x55, 0x5E, 0x59, 0x20, 0x1E, /* 216-223 */
+    0x7F, 0x61, 0x61, 0x61, 0x7B, 0x0F, 0x1D, 0x63, /* 224-231: ..ä, å... */
+    0x04, 0x05, 0x65, 0x65, 0x07, 0x69, 0x69, 0x69, /* 232-239 */
+    0x20, 0x7D, 0x08, 0x6F, 0x6F, 0x6F, 0x7C, 0x2F, /* 240-247: ..ö... */
+    0x0C, 0x06, 0x75, 0x75, 0x7E, 0x79, 0x20, 0x79, /* 248-255 */
+];
+
+// Masks need when encoding ASCII to GSM 7 bit
+const STR_7BIT_ESCAPE_IND: u8 = 0x80;
+const STR_7BIT_CODE_MASK: u8 = 0x7F;
+const STR_7BIT_ESCAPE_CODE: u8 = 0x1B;
+
+/// A struct holding both number and message with can be send as an SMS
+pub struct Sms<'a> {
+    number: &'a str,
+    message: &'a str,
+}
+
+impl<'a> Sms<'a> {
+    /// Creates a new Sms message
+    /// `number` should be in national format, including the country code at start. The + character is not need at start and will be ignored.
+    /// Max `message` lenght 160 chars
+    pub fn new(number: &'a str, message: &'a str) -> Self {
+        Self { number, message }
+    }
+    // Encode number in the way modem expect it
+    // Reimplement from https://github.com/nrfconnect/sdk-nrf/blob/main/lib/sms/sms_submit.c#L46
+    fn encode_number(number: &str) -> Result<String<15>, Error> {
+        let mut encoded_number = String::from(number.trim_start_matches("+"));
+
+        if encoded_number.len() % 2 != 0 {
+            encoded_number
+                .push('F')
+                .map_err(|_| Error::BufferTooSmall(None))?;
+        }
+
+        if number.is_ascii() {
+            // Since we are checking if the number of characters is even before this unsafe
+            // and we only allow ASCII chars doing this swap shouldn't have any UB
+            unsafe {
+                for c in encoded_number.as_bytes_mut().array_chunks_mut::<2>() {
+                    c.swap(0, 1);
+                }
+            }
```

Comment:
Is unsafe really necessary here? If you do to-bytes, swap, and then use from-bytes, I would guess that the optimizations will make it in place.

Also, do we really need the array chunks here? Would the normal chunks not be able to do this? It would save a nightly feature dependency.

Reply:
I think I found a way to do it without array_chunks and without unsafe.
It is slight hacky because I only find a way to go through ArrayVec. ArrayString doesn't have an to_bytes function, only as_bytes.